tf_quant_finance/black_scholes/approximations/american_option.py

```diff
@@ -343,6 +343,210 @@ def _calc_q(n, m, sign, k=1):
+      discount_factors = tf.exp(-discount_rates * expiries)
+    elif discount_factors is not None:
+      discount_factors = tf.convert_to_tensor(
+          discount_factors, dtype=dtype, name='discount_factors')
+      discount_rates = -tf.math.log(discount_factors) / expiries
+    else:
+      discount_rates = tf.convert_to_tensor(
+          0.0, dtype=dtype, name='discount_rates')
+      discount_factors = tf.convert_to_tensor(
+          1.0, dtype=dtype, name='discount_factors')
+
+    if dividend_rates is None:
+      dividend_rates = tf.convert_to_tensor(
+          0.0, dtype=dtype, name='dividend_rates')
+
+    if cost_of_carries is not None:
+      cost_of_carries = tf.convert_to_tensor(
+          cost_of_carries, dtype=dtype, name='cost_of_carries')
+    else:
+      cost_of_carries = discount_rates - dividend_rates
+
+    if forwards is not None:
+      forwards = tf.convert_to_tensor(forwards, dtype=dtype, name='forwards')
+    else:
+      spots = tf.convert_to_tensor(spots, dtype=dtype, name='spots')
+      forwards = spots * tf.exp(cost_of_carries * expiries)
+
+    # If cost of carry is greater than or equal to discount rate, then use Black-Scholes price
+    american_prices = tf.where(cost_of_carries >= discount_rates and is_call_options,
+            vanilla_prices.option_price(
+                volatilities=volatilities,
+                strikes=strikes,
+                expiries=expiries,
+                forwards=forwards,
+                discount_rates=discount_rates,
+                cost_of_carries=cost_of_carries,
+                is_call_options=is_call_options),
+            tf.where(is_call_options,
+                _call_1993(forwards, strikes, expiries, discount_rates, cost_of_carries, volatilities),
+                _call_1993(strikes, forwards, expiries, discount_rates - cost_of_carries, -cost_of_carries, volatilities)))
+
+    return american_prices
+
+
+def _call_1993(S, X, T, r, b, sigma):
+  beta = (0.5 - b / sigma**2) + tf.math.sqrt((b / sigma**2 - 0.5)**2 + 2 * r / sigma**2)
+  alpha = (_boundary_1993(X, T, r, b, sigma, beta) - X) * (_boundary_1993(X, T, r, b, sigma, beta))**(-beta)
+
+  return tf.where(S >= _boundary_1993(X, T, r, b, sigma, beta),
+          S - X,
+          alpha * S**beta -
+          alpha * _phi_1993(S, T, beta, _boundary_1993(X, T, r, b, sigma, beta), _boundary_1993(X, T, r, b, sigma, beta), r, b, sigma) +
+          _phi_1993(S, T, 1, _boundary_1993(X, T, r, b, sigma, beta), _boundary_1993(X, T, r, b, sigma, beta), r, b, sigma) -
+          _phi_1993(S, T, 1, X, _boundary_1993(X, T, r, b, sigma, beta), r, b, sigma) - 
+          X * _phi_1993(S, T, 0, _boundary_1993(X, T, r, b, sigma, beta), _boundary_1993(X, T, r, b, sigma, beta), r, b, sigma) +
+          X * _phi_1993(S, T, 0, X, _boundary_1993(X, T, r, b, sigma, beta), r, b, sigma))
+
+
+def _phi_1993(S, T, gamma, H, I, r, b, sigma):
+  kappa = (2 * b) / sigma**2 + (2 * gamma - 1)
```

Comment:
please add docstring stating that this corresponds to `phi` (7) from the paper

Reply:
Agreed.